prov/verbs_ud/src/verbs_ud.h

```diff
@@ -0,0 +1,738 @@
+	.comp_order = FI_ORDER_STRICT,				\
+	.iov_limit = VERBS_UD_MSG_IOV_LIMIT,			\
+	.rma_iov_limit = VERBS_UD_MSG_IOV_LIMIT			\
+};								\
+								\
+/* Notes: */							\
+/* - size, iov_limit are assigned at run-time */		\
+/*   in @ofi_verbs_ud_info_init */				\
+static struct fi_rx_attr ofi_verbs_ud_rx_attr = {		\
+	.caps = FI_MSG | FI_RECV,				\
+	.comp_order = FI_ORDER_STRICT,				\
+	.iov_limit = VERBS_UD_MSG_IOV_LIMIT			\
+};								\
+								\
+/* Notes: */							\
+/* - name is assigned at run-time in @ofi_verbs_ud_info_init */	\
+static struct fi_fabric_attr ofi_verbs_ud_fabric_attr = {	\
+	.prov_version = FI_VERSION(OFI_VERBS_UD_MAJOR_VERSION,	\
+				   OFI_VERBS_UD_MINOR_VERSION)	\
+};								\
+								\
+/* Notes: */							\
+/* - cq_cnt, ep_cnt, max_ep_tx|rx_ctx, tx|rx_ctx_cnt */		\
+/*   are assigned at run-time in @ofi_verbs_ud_info_init */	\
+static struct fi_domain_attr ofi_verbs_ud_domain_attr = {	\
+	.threading		= FI_THREAD_SAFE,		\
+	.control_progress	= FI_PROGRESS_AUTO,		\
+	.data_progress		= FI_PROGRESS_AUTO,		\
+	.resource_mgmt		= FI_RM_DISABLED,		\
+	.mr_mode		= OFI_MR_BASIC_MAP |		\
+				  FI_MR_LOCAL | FI_MR_BASIC,	\
+	.mr_key_size		= sizeof_field(struct ibv_sge,	\
+					       lkey),		\
+	.cq_data_size		= sizeof_field(			\
+					struct ibv_send_wr,	\
+					imm_data),		\
+	.tx_ctx_cnt		= 1024, /* TODO: const */	\
+	.rx_ctx_cnt		= 1024,				\
+	.mr_iov_limit		= VERBS_UD_MSG_IOV_LIMIT,	\
+	/* max_err_data is size of ibv_wc::vendor_err */	\
+	/*for CQ, 0 - for EQ */					\
+	.max_err_data		= sizeof_field(struct ibv_wc,	\
+					       vendor_err),	\
+};								\
+								\
+/* Notes: */							\
+/* - max_msg_size, max_order_raw_size, max_order_waw_size, */	\
+/*   protocol */						\
+/*   are assigned at run-time in @ofi_verbs_ud_info_init */	\
+static struct fi_ep_attr ofi_verbs_ud_ep_attr = {		\
+	.type = FI_EP_DGRAM,					\
+	.protocol = FI_PROTO_IB_UD,				\
+	.protocol_version = 0					\
+}
+
+#define VERBS_UD_INJECT_FLAGS(ep, len, flags)			\
+	(((flags & FI_INJECT) ||				\
+	  (len <= ep->info->tx_attr->inject_size)) ?		\
+	 IBV_SEND_INLINE : 0)
+#define VERBS_UD_INJECT(ep, len)				\
+	VERBS__UD_INJECT_FLAGS(ep, len,				\
+			       ep->info->tx_attr->op_flags)
+
+#define VERBS_UD_SELECTIVE_COMP(ep)				\
+	(ep->ep_flags & FI_SELECTIVE_COMPLETION)
+#define VERBS_UD_COMP_FLAGS(ep, flags)				\
+	((!VERBS_UD_SELECTIVE_COMP(ep) ||			\
+	  (flags & (FI_COMPLETION | FI_TRANSMIT_COMPLETE))) ?	\
+	 IBV_SEND_SIGNALED : 0)
+#define VERBS_UD_COMP(ep)					\
+	VERBS_UD_COMP_FLAGS(ep, ep->info->tx_attr->op_flags)
+
+#define OFI_IBV_MTU_TYPE(bytes, ret_bytes)		\
+	case IBV_MTU_##bytes:			\
+		ret_bytes = bytes;		\
+		break
+
+static inline
+size_t ofi_verbs_ud_mtu_type_to_len(enum ibv_mtu mtu_type)
+{
+	size_t bytes = 0;
+
+	switch (mtu_type) {
+	OFI_IBV_MTU_TYPE(256, bytes);
+	OFI_IBV_MTU_TYPE(512, bytes);
+	OFI_IBV_MTU_TYPE(1024, bytes);
+	OFI_IBV_MTU_TYPE(2048, bytes);
+	OFI_IBV_MTU_TYPE(4096, bytes);
```

Comment:
I think it would be just as clean to drop the macro and open-code this, adding a default case.  I would move this into a C file, rather than forcing static inline.